src/ray/gcs/gcs_server/gcs_server_main.cc

```diff
@@ -54,7 +55,16 @@ int main(int argc, char *argv[]) {
 
   RayConfig::instance().initialize(config_map);
 
-  boost::asio::io_service main_service;
+  size_t io_service_num_{2};
+  std::shared_ptr<ray::IOServicePool> io_service_pool =
+      std::make_shared<ray::IOServicePool>(io_service_num_);
+  io_service_pool->Run();
+
+  // IO Service for signals handler.
+  boost::asio::io_service signal_service;
```

Comment:
Can we just use a main thread as a main_service (and don't have a separate signal service) & just an additional io service only for node_manager? It seems a little weird we have io service that is only for signal.

Reply:
I've tried reuse main_service or one of main_services for signal handler but there will be a deadlock since we join the threads in signal handler, which also uses the thread which will be joined.

Reply:
I probably didn't understand here. If you use main_service on the main thread, there's no need to join them right? What I am saying here is 

```
boost::asio::io_service main_service;
...
boost::asio::signal_set signals(main_service);
```

That says, we just need to create io_service_pool with value 1. (for node manager heartbeat handler) & just have another main_service on the main thread for both main service & signal service

Reply:
EX) https://github.com/ray-project/ray/blob/d4324a4a8f886f60cd5f12356b97466e86a78b89/src/ray/raylet/main.cc#L246

Reply:
In conclusion, we will only have 2 threads, main thread with main_service that is in charge of both signal & main logic + extra thread that is dedicating to node manager detctor

Reply:
> I probably didn't understand here. If you use main_service on the main thread, there's no need to join them right? What I am saying here is
> 
> ```
> boost::asio::io_service main_service;
> ...
> boost::asio::signal_set signals(main_service);
> ```
> 
> That says, we just need to create io_service_pool with value 1. (for node manager heartbeat handler) & just have another main_service on the main thread for both main service & signal service

The thread will be join in stop method  of  io_service_pool.

Reply:
In my example, it doesn't matter that `io_service_pool` joins inside signal handler. What I am saying is just to define 1 thread inside `io_service_pool`, which is just for NodeFailureDetector. And just keep the main_service for both signal & main thread logic. It should not cause deadlocks.

Reply:
```cpp
// Use 1 thread for node failure detector event loop.
io_service_pool_ = std::make_shared<IOServicePool>(1);

// Share main_service for both signal & main event loop
boost::asio::io_service main_service;
...
boost::asio::signal_set signals(main_service);
```

Reply:
> ```c++
> // Use 1 thread for node failure detector event loop.
> io_service_pool_ = std::make_shared<IOServicePool>(1);
> 
> // Share main_service for both signal & main event loop
> boost::asio::io_service main_service;
> ...
> boost::asio::signal_set signals(main_service);
> ```

This solution will work, but why don't we separate the main service and signal service but merge them together?

Reply:
I think there's no need to add extra complexity (I think additional threads are always adding complexity) when it is not necessary.

Reply:
@rkooo567 reasonable. Ockham's Razor rule. :)